fuel/datasets/hdf5.py

```diff
@@ -316,11 +322,21 @@ def _out_of_memory_get_data(self, state=None, request=None):
             if isinstance(request, slice):
                 request = slice(request.start + subset.start,
                                 request.stop + subset.start, request.step)
+                data = state[source_name][request]
             elif isinstance(request, list):
                 request = [index + subset.start for index in request]
+                try:
+                    data = state[source_name][request]
+                except TypeError:
```

Comment:
Right. The problem is that at this level in the code `H5PYDataset` can't know for sure whether two consecutive calls of `get_data` originate from the same data stream.

We could tie the `sort_needed` flag to a specific `state` argument, but we would need to compare objects using the `is` operator instead of `==`, as different `h5py.File` instances with the same path/`FileID` compare equal.